lib/net.c

```diff
@@ -41,18 +43,53 @@
 #define GATEKEEPER_PKT_DROP_QUEUE (127)
 
 static struct net_config config;
-
 /*
- * XXX The secret key of the RSS hash must be random
- * in order to avoid hackers to know it.
+ * The secret key of the RSS hash (RSK) must be random in order
+ * to prevent hackers from knowing it.
  */
-uint8_t default_rss_key[GATEKEEPER_RSS_KEY_LEN] = {
-	0x6d, 0x5a, 0x56, 0xda, 0x25, 0x5b, 0x0e, 0xc2,
-	0x41, 0x67, 0x25, 0x3d, 0x43, 0xa3, 0x8f, 0xb0,
-	0xd0, 0xca, 0x2b, 0xcb, 0xae, 0x7b, 0x30, 0xb4,
-	0x77, 0xcb, 0x2d, 0xa3, 0x80, 0x30, 0xf2, 0x0c,
-	0x6a, 0x42, 0xb7, 0x3b, 0xbe, 0xac, 0x01, 0xfa,
-};
+uint8_t default_rss_key[GATEKEEPER_RSS_KEY_LEN];
+
+static int
+randomize_rss_key(void)
+{
+	uint16_t final_set_count;
+	int ret;
+
+	/*
+	 * To validate if the key generated is reasonable, the
+	 * number of bits set to 1 in the key must be greater than 
+	 * 10% and less than 90% of the total bits in the key.
+	 * min_num_set_bits and max_num_set_bits represent the lower 
+	 * and upper bound for the key.
+	 */
+	const uint16_t min_num_set_bits = sizeof(default_rss_key) * 8 * 0.1;
+	const uint16_t max_num_set_bits = sizeof(default_rss_key) * 8 * 0.9;
+
+	int number_of_bytes = 0;
+	do {
+		/* 
+		 * The last parameter to syscall getrandom i.e flags will alter
+		 * the behaviour of the call. In the case where flag == 0, the call
+		 * will block until the pool has the required entropy and uses the
+		 * /dev/urandom pool. The GRND_RANDOM flag bit can be used to
+		 * switch to the /dev/random pool.
+		 */	
+		do {
+			ret = syscall(SYS_getrandom, default_rss_key + number_of_bytes,
+					sizeof(default_rss_key) - number_of_bytes, 0);
+			if (ret < 0)
+				return -1;
+			number_of_bytes += ret;	
+		} while (number_of_bytes < sizeof(default_rss_key))
+
+		uint8_t i;
```

Comment:
Move the declaration of variable `i` to the beginning of the block, that is, on the line just after `do {`.